test/clj/rems/db/test_transactions.clj

```diff
@@ -0,0 +1,177 @@
+                                                     (fn [] (sample-until-interrupted
+                                                             (fn [] (read-app-events app-id))))))
+        all-events-readers (submit-all thread-pool (for [_ (range concurrent-readers)]
+                                                     (fn [] (sample-until-interrupted
+                                                             (fn [] (read-all-events))))))
+        writers (submit-all thread-pool (for [app-id app-ids
+                                              _ (range writers-per-application)]
+                                          (fn [] (sample-until-interrupted
+                                                  (fn [] (write-event app-id))))))]
+    (Thread/sleep test-duration-millis)
+    (doto thread-pool
+      (.shutdownNow)
+      (.awaitTermination 30 TimeUnit/SECONDS))
+
+    (let [writer-results (->> (flatten (map #(.get ^Future %) writers))
+                              (remove #{::transaction-conflict}))
+          app-events-reader-results (flatten (map #(.get ^Future %) app-events-readers))
+          all-events-reader-results (flatten (map #(.get ^Future %) all-events-readers))
+          final-events (applications/get-all-events-since 0)
+          final-events-by-app-id (group-by :application/id final-events)]
+
+      (testing "all commands succeeded"
+        (is (not (empty? writer-results)))
+        (is (every? nil? writer-results))) ; successful commands return nil
+
+      (testing "all events were written"
+        (is (= (+ (count app-ids) ; one application created event per application
+                  (count writer-results))
+               (count final-events))))
+
+      (testing "event IDs are observed in monotonically increasing order, within one application"
+        (doseq [observed app-events-reader-results]
+          (let [final-events (get final-events-by-app-id (::app-id observed))]
+            (is (= (->> final-events
+                        (take (count (::events observed)))
+                        (map :event/id))
+                   (->> (::events observed)
+                        (map :event/id)))))))
+
+      (testing "event IDs are observed in monotonically increasing order, globally"
+        (doseq [observed all-events-reader-results]
+          (is (= (->> final-events
+                      (take (count (::events observed)))
+                      (map :event/id))
+                 (->> (::events observed)
+                      (map :event/id))))))
+
+      (testing "commands are executed serially"
+        (doseq [[_ events] final-events-by-app-id]
+          (is (= (->> (range 1 (count events))
+                      (map str))
+                 (->> events
+                      (filter #(= :application.event/draft-saved (:event/type %)))
+                      (map #(get-in % [:application/field-values observed-app-version-marker])))))))
+
+      ;; TODO?
```

Comment:
please clarify. this does not pass right now due to transaction rollbacks, right?

Reply:
Yes, transaction rollbacks. To guarantee that there are no gaps, it would be possible to use `max(id) + 1`.